stan/math/rev/core/operator_subtraction.hpp

```diff
@@ -133,7 +107,212 @@ inline var operator-(const var& a, Arith b) {
  */
 template <typename Arith, require_arithmetic_t<Arith>* = nullptr>
 inline var operator-(Arith a, const var& b) {
-  return {new internal::subtract_dv_vari(a, b.vi_)};
+  return make_callback_vari(a - b.vi_->val_,
+                            [bvi = b.vi_, a](const auto& vi) mutable {
+                              if (unlikely(is_any_nan(a, bvi->val_))) {
+                                bvi->adj_ = NOT_A_NUMBER;
+                              } else {
+                                bvi->adj_ -= vi.adj_;
+                              }
+                            });
+}
+
+/**
+ * Subtraction operator for matrix variables (C++).
+ *
+ * @tparam VarMat1 A matrix of vars or a var with an underlying matrix type.
+ * @tparam VarMat2 A matrix of vars or a var with an underlying matrix type.
+ * @param a First variable operand.
+ * @param b Second variable operand.
+ * @return Variable result of subtracting two variables.
+ */
+template <typename VarMat1, typename VarMat2,
+          require_all_rev_matrix_t<VarMat1, VarMat2>* = nullptr>
+inline auto subtract(const VarMat1& a, const VarMat2& b) {
+  check_matching_dims("subtract", "a", a, "b", b);
+  using op_ret_type = decltype(a.val() - b.val());
+  using ret_type = promote_var_matrix_t<op_ret_type, VarMat1, VarMat2>;
+  arena_t<ret_type> ret((a.val() - b.val()));
+  arena_t<VarMat1> arena_a = a;
+  arena_t<VarMat2> arena_b = b;
+  reverse_pass_callback([ret, arena_a, arena_b]() mutable {
```

Comment:
Why do we want to get rid of `reverse_pass_callback()`? Is there some efficiency gain? tbh I like how it looks and there are times that we want to have different callbacks go off for different inputs. Would this looks like 

```cpp
return ret_type(make_callback_vari(arena_a.val() - arena_b.val(), 
[arena_a, arena_b](const auto& vi) mutable {
    for (Eigen::Index i = 0; i < vi.size(); ++i) {
      const auto ret_adj = vi.adj_.coeffRef(i);
      arena_a.adj().coeffRef(i) += ret_adj;
      arena_b.adj().coeffRef(i) -= ret_adj;
    }
  }));
```

Reply:
I'm okay with that, though I like how reverse_pass_callback perpetrates out the reverse and forward passes

Reply:
There is efficiency gain,, as I demonstrated with scalar addition. Most likely it will also have non-negligible gain for small vectors. Also `reverse_pass_callback` and `make_callback_vari` are solving the same issue and I don't like duplicated functionality.

> there are times that we want to have different callbacks go off for different inputs.

You can also do that with `make_callback_vari`.

Reply:
Aye, looking at `make_callback_vari` again I need to switch.

Was there a reason we didn't do `make_callback_var`? It looks like this could make the return type algebra a lot easier.

I think `reverse_pass_callback` probably has some uses? Like `make_callback_vari` doesn't handle `std::vector` yet (unless I'm reading somethign wrong)

Reply:
> Like make_callback_vari doesn't handle std::vector yet

It should handle `std::vector`s as well. Although I did forget to add a test for `std::vector`. I will open a PR soon.

Reply:
No, actually you are right. `make_callback_vari` can only handle types that are supported in `vari_value`.

Reply:
Does `make_callback_vari()` support `Eigen::Matrix<var>`? Looking at `callback_vari` I'm seeing

```cpp
template <typename T, typename F>
struct callback_vari : public vari_value<T> {
  F rev_functor_;

  explicit callback_vari(T&& value, F&& rev_functor)
      : vari_value<T>(std::move(value)),
        rev_functor_(std::forward<F>(rev_functor)) {}

  inline void chain() final { rev_functor_(*this); }
};
```

So for like

```cpp
template <typename VarMat1, typename VarMat2,
          require_all_rev_matrix_t<VarMat1, VarMat2>* = nullptr>
inline auto add(const VarMat1& a, const VarMat2& b) {
  check_matching_dims("add", "a", a, "b", b);
  using op_ret_type = decltype(a.val() + b.val());
  using ret_type = promote_var_matrix_t<op_ret_type, VarMat1, VarMat2>;
  arena_t<VarMat1> arena_a(a);
  arena_t<VarMat2> arena_b(b);
  return ret_type(make_callback_vari(arena_a.val() + arena_b.val(),
   [arena_a, arena_b](const auto& vi) mutable {
     for (Eigen::Index i = 0; i < ret.size(); ++i) {
       const auto ref_adj = this->adj_.coeffRef(i);
       arena_a.adj().coeffRef(i) += ref_adj;
       arena_b.adj().coeffRef(i) += ref_adj;
     }
   }));
}
```

So for here when both are eigen matrices of vars, the inner type `T` is an `Eigen::Matrix<var, -1, -1>`. But then `callback_vari` will inherit from `var_value<Matrix<var>>` right?

Reply:
`callback_vari` is a single vari. So you can't use it to to return whole matrix. However, you could make a matrix and use `make_callback_vari` for each scalar in the matrix.

Reply:
icic. Okay then for this PR I think I'd like to keep using `reverse_pass_callback()` as is since I need it to work with `var<mat>` and `mat<var>`

Reply:
Yeah I think `callback_vari` is going to be annoying to work with for `mat<var>` and `std::vector<var>` types. I'd rather just roll with `reverse_pass_callback` for those and leave `make_callback_vari` for `var_value<Eigen>` types.

A matrix of callback varis would also push a ton of varis into the chaining stack (each of those being virtual function calls). The way Steve has done it here is just one virtual function call, so I expect this to be faster even if it makes an extra vari.